<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Documentation - dna-parser</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Documentation";
        var mkdocs_page_input_path = "documentation.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> dna-parser
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Documentation</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#encodings">Encodings</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#atomic-number">Atomic Number</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#chaos-game">Chaos Game</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_1">Function Arguments:</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_1">Output:</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cross">Cross</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_2">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_2">Output:</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#dna-walk">DNA Walk</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_3">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_3">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#eiip">EIIP</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_4">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_4">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#fickett-score">Fickett Score</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_5">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_5">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#onehot-voss">Onehot (Voss)</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_6">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_6">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#real-number-or-pam">Real-number (or PAM)</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_7">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_7">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tf-idf">TF-IDF</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#methods">Methods:</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#attributes">Attributes</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#examples">Examples</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#z-curve">Z-Curve</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_8">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_8">Output</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#importing-sequences">Importing Sequences</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#importing-fasta-files">Importing Fasta Files</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_9">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_9">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#importing-sequences-only">Importing Sequences Only</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_10">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_10">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#importing-metadata-only">Importing Metadata Only</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_11">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_11">Output</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#other-functions">Other Functions</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#kmers">Kmers</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_12">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_12">Output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#generating-random-sequences">Generating Random Sequences</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#function-arguments_13">Function Arguments</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#output_13">Output</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../about/">About</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../license/">License</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">dna-parser</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Documentation</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation">Documentation</h1>
<h2 id="encodings">Encodings</h2>
<h3 id="atomic-number">Atomic Number</h3>
<pre><code class="language-python">dna_parser.atomic_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences)</strong>. Each nucleotide is encoded as its atomic number:</p>
<ul>
<li>A= 70</li>
<li>C= 58</li>
<li>G= 78</li>
<li>T/U= 66</li>
<li>Other characters or gaps = 0</li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.atomic_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[[70 78 66]
# [70 58 58]]
#
# (2, 3)
</code></pre>
<p><br/></p>
<h3 id="chaos-game">Chaos Game</h3>
<pre><code class="language-python">dna_parser.chaos_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_1">Function Arguments:</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_1">Output:</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences, 2)</strong>. <br />Each sequence is encoded in a square with vertices A: (1,1), C: (-1,-1),
G: (1,-1), T/U: (-1,1). The sequence representation starts at the center of the square in (0,0). The first nucleotide is represented as a point halfway between the starting point and its corresponding vertice. 
Each following nucleotide a new point halfaway between the previous point and its corresponding vertice. If a character other than A,C,G,T or U is encountered, the values are not updated and values from the previous point are used.</p>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.chaos_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#     x  ,   y
#[[[ 0.5    0.5  ]
#  [ 0.75  -0.25 ]
#  [-0.125  0.375]]
#
# [[ 0.5    0.5  ]
#  [-0.25  -0.25 ]
#  [-0.625 -0.625]]]
#(2, 3, 2)
</code></pre>
<p><br/></p>
<h3 id="cross">Cross</h3>
<pre><code class="language-python">dna_parser.cross_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_2">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_2">Output:</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences,2).</strong> Each nucleotide is encoded as follows:</p>
<ul>
<li>A= [0,-1]</li>
<li>C= [-1,0]</li>
<li>G= [1,0]</li>
<li>T/U= [0,1]</li>
<li>Other characters or gaps = [0,0]</li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.cross_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[[[ 1  1]
#  [-1  1]
#  [ 1 -1]]
#
# [[ 1  1]
#  [-1 -1]
#  [-1 -1]]]
#
#(2, 3, 2)

</code></pre>
<p><br/></p>
<h3 id="dna-walk">DNA Walk</h3>
<pre><code class="language-python">dna_parser.dna_walk(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_3">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_3">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences, 2).</strong> <br/> Each sequence is represented on a 2D grid, with its representation starting at coordinates (0,0). For each nucleotide in the sequence coordinates are updated to form a path as follows:</p>
<ul>
<li>A: <span class="arithmatex"><span class="MathJax_Preview">x_{n+1}= x_{n}-1</span><script type="math/tex">x_{n+1}= x_{n}-1</script></span>; &nbsp; <span class="arithmatex"><span class="MathJax_Preview">y_{n+1}= y_{n}</span><script type="math/tex">y_{n+1}= y_{n}</script></span></li>
<li>C: <span class="arithmatex"><span class="MathJax_Preview">x_{n+1}= x_{n}</span><script type="math/tex">x_{n+1}= x_{n}</script></span>; &nbsp; <span class="arithmatex"><span class="MathJax_Preview">y_{n+1}= y_{n}-1</span><script type="math/tex">y_{n+1}= y_{n}-1</script></span></li>
<li>G: <span class="arithmatex"><span class="MathJax_Preview">x_{n+1}= x_{n}</span><script type="math/tex">x_{n+1}= x_{n}</script></span>; &nbsp; <span class="arithmatex"><span class="MathJax_Preview">y_{n+1}= y_{n}+1</span><script type="math/tex">y_{n+1}= y_{n}+1</script></span></li>
<li>T/U: <span class="arithmatex"><span class="MathJax_Preview">x_{n+1}= x+1</span><script type="math/tex">x_{n+1}= x+1</script></span>; &nbsp; <span class="arithmatex"><span class="MathJax_Preview">y_{n+1}= y_{n}</span><script type="math/tex">y_{n+1}= y_{n}</script></span></li>
<li>Other characters or gaps: <span class="arithmatex"><span class="MathJax_Preview">x_{n+1}= x_{n}</span><script type="math/tex">x_{n+1}= x_{n}</script></span>; &nbsp; <span class="arithmatex"><span class="MathJax_Preview">y_{n+1}= y_{n}</span><script type="math/tex">y_{n+1}= y_{n}</script></span></li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.dna_walk(sequences)
print(encoding)
print(encoding.shape)

# Output:
#   x , y
#[[[-1  0]
#  [-1  1]
#  [ 0  1]]
#
# [[-1  0]
#  [-1 -1]
#  [-1 -2]]]
#(2, 3, 2)

</code></pre>
<p><br/></p>
<h3 id="eiip">EIIP</h3>
<pre><code class="language-python">dna_parser.eiip_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_4">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_4">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences).</strong> Each nucleotide is encoded as its electron-ion interaction pseudopotential (EIIP):</p>
<ul>
<li>A= 0.1260</li>
<li>C= 0.1340</li>
<li>G= 0.0806</li>
<li>T/U= 0.1335</li>
<li>Other characters or gaps = 0.0</li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.eiip_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[[0.126  0.0806 0.1335]
# [0.126  0.134  0.134 ]]
#
# (2, 3)
</code></pre>
<p><br/></p>
<h3 id="fickett-score">Fickett Score</h3>
<pre><code class="language-python">dna_parser.fickett_score(sequences, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_5">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_5">Output</h4>
<p><strong>Numpy array with shape (number of sequences).</strong> <br/>Compute the probability of each sequence to be a coding sequence. See the <a href="../about/#fickett-score">About</a> section for more details.</p>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.fickett_score(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[0.3203 0.407 ]
#
#(2,)
</code></pre>
<p><br/></p>
<h3 id="onehot-voss">Onehot (Voss)</h3>
<pre><code class="language-python">dna_parser.onehot_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_6">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_6">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences, 4).</strong> Each nucleotide is encoded as follows:</p>
<ul>
<li>C= [1,0,0,0]</li>
<li>G= [0,1,0,0]</li>
<li>A= [0,0,1,0]</li>
<li>T/U= [0,0,0,1]</li>
<li>Other characters or gaps = [0,0,0,0]</li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.onehot_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[[[0 0 1 0]
#  [0 1 0 0]
#  [0 0 0 1]]

# [[0 0 1 0]
#  [1 0 0 0]
#  [1 0 0 0]]]

# (2, 3, 4)
</code></pre>
<p><br/></p>
<h3 id="real-number-or-pam">Real-number (or PAM)</h3>
<pre><code class="language-python">dna_parser.real_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_7">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_7">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences).</strong> Each nucleotide is encoded as follows:</p>
<ul>
<li>A= -1.5</li>
<li>G= -0.5</li>
<li>C= 0.5</li>
<li>T/U= 1.5</li>
<li>Other characters or gaps = 0</li>
</ul>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agt&quot;,&quot;acc&quot;]
encoding= dps.real_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#[-1.5 -0.5  1.5]
#[-1.5  0.5  0.5]]
#
# (2, 3)
</code></pre>
<p><br/></p>
<h3 id="tf-idf">TF-IDF</h3>
<pre><code class="language-python">#Class
dna_parser.Tfidf(corpus , kmer, vocabulary= None)
</code></pre>
<p><strong>Parameters:</strong> </p>
<ul>
<li><strong>corpus</strong> (list of str): list of genomic sequences.</li>
<li><strong>kmer</strong> (int): length to use to generate kmers in the sequences.</li>
<li><strong>vocabulary</strong> (dict(str:int)): Dictionary mapping each kmers to consider for encoding to a unique integer value.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available. 
<br/></li>
</ul>
<h4 id="methods">Methods:</h4>
<h5 id="add_to_corpus">add_to_corpus</h5>
<pre><code class="language-python">Tfidf.add_to_corpus(new_corpus)
</code></pre>
<ul>
<li><strong>new_corpus</strong>: (list of str): list of genomic sequences.
Adds sequences to the existing corpus.</li>
</ul>
<h5 id="fit">fit</h5>
<pre><code class="language-python">Tfidf.fit()
</code></pre>
<p>Fits the Tfidf instance. Compiles the vocabulary if it is not provided. Computes the Inverse Document Frequency.</p>
<h5 id="fit_transform">fit_transform</h5>
<pre><code class="language-python">Tfidf.fit_transform(sequences= None, normalization= &quot;L2&quot;)
</code></pre>
<ul>
<li><strong>sequences</strong> (list of str or None): list of genomic sequences to transform.</li>
<li><strong>normalization</strong> (str): "L2" for L2 normalization. Anything else results in no normalization.</li>
</ul>
<p>Fits the Tfidf instance. Compiles the vocabulary if it is not provided. Computes the Inverse Document Frequency
and transforms the sequences in their TF-IDF representation. If "sequences= None", transforms the corpus.</p>
<h5 id="set_threads">set_threads</h5>
<pre><code class="language-python">Tfidf.set_threads(n_jobs)
</code></pre>
<p><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</p>
<p>Adjusts the number of threads used to encode the sequences in parallel.</p>
<h5 id="set_vocabulary">set_vocabulary</h5>
<pre><code class="language-python">Tfidf.set_vocabulary(vocabulary):
</code></pre>
<ul>
<li><strong>vocabulary</strong>: (dict(str:int)) Dictionary mapping each kmers to consider for encoding to a unique integer value.</li>
</ul>
<p>The integer associated with each kmer needs to be unique, continuous, and start at 0.</p>
<pre><code class="language-python">#this vocabulary is correct:
{&quot;gtc&quot;:0, &quot;atg&quot;:1, &quot;acg&quot;:2}

#these are not correct:
{&quot;gtc&quot;:1, &quot;atg&quot;:2, &quot;acg&quot;:3}
{&quot;gtc&quot;:0, &quot;atg&quot;:2, &quot;acg&quot;:3}
</code></pre>
<h5 id="transform">transform</h5>
<pre><code class="language-python">Tfidf.transform(sequences= None, normalization= &quot;L2&quot;)
</code></pre>
<ul>
<li><strong>sequences</strong>: (list of str or None): list of genomic sequences to transform.</li>
<li><strong>normalization</strong>: (str): "L2" for L2 normalization. Anything else results in no normalization.</li>
</ul>
<p>Transforms the sequences in their TF-IDF representation. If "sequences= None", transforms the corpus. The Tfidf instance
needs to be fitted with the fit() or fit_transform() function before calling transform().</p>
<h4 id="attributes">Attributes</h4>
<pre><code class="language-python">Tfidf.vocabulary # None or dict(str:int)
</code></pre>
<pre><code class="language-python">Tfidf.corpus     # List(str)
</code></pre>
<pre><code class="language-python">Tfidf.kmer_size  # Int
</code></pre>
<pre><code class="language-python">Tfidf.idf        # None or numpy array
</code></pre>
<pre><code class="language-python">Tfidf.is_idf_uptodate # Bool
</code></pre>
<pre><code class="language-python">Tfidf.n_jobs # Int
</code></pre>
<h4 id="examples">Examples</h4>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agtcgc&quot;,&quot;accgtc&quot;]
tfidf= dps.Tfidf(sequences,2)
tfidf.fit()
encoding= tfidf.transform()
print(encoding)
print(encoding.shape)

# Output:
#&lt;Compressed Sparse Row sparse matrix of dtype 'float64'
#        with 2 stored elements and shape (2, 5)&gt;
#  Coords        Values
#  (0, 1)        -0.1351550360360548
#  (1, 1)        -0.1351550360360548
#
# (2, 5)
</code></pre>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;attcggagt&quot;,&quot;attctggga&quot;]
tfidf= dps.Tfidf(sequences,3)
encodings= tfidf.fit_transform()

tfidf.add_to_corpus([&quot;agccgcgga&quot;])
encodings2= tfidf.fit_transform(normalization= None)
print(encodings)
print(encodings2)

# Output:
#&lt;Compressed Sparse Row sparse matrix of dtype 'float64'
#        with 6 stored elements and shape (2, 5)&gt;
#  Coords        Values
#  (0, 0)        0.0
#  (0, 1)        0.7071067811865476
#  (0, 2)        0.7071067811865476
#  (1, 0)        0.0
#  (1, 3)        0.7071067811865476
#  (1, 4)        0.7071067811865476
#&lt;Compressed Sparse Row sparse matrix of dtype 'float64'
#        with 7 stored elements and shape (3, 5)&gt;
#  Coords        Values
#  (0, 0)        0.4054651081081644
#  (0, 1)        1.0986122886681098
#  (0, 2)        1.0986122886681098
#  (1, 0)        0.4054651081081644
#  (1, 3)        1.0986122886681098
#  (1, 4)        0.4054651081081644
#  (2, 4)        0.4054651081081644
</code></pre>
<p><br/></p>
<h3 id="z-curve">Z-Curve</h3>
<pre><code class="language-python">dna_parser.zcurve_encoding(sequences, pad_type= &quot;after&quot;, pad_length= -2, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_8">Function Arguments</h4>
<ul>
<li><strong>sequences</strong> (list of str): list of genomic sequences.</li>
<li><strong>pad_type</strong> (str): pad (or trim) "before" or "after" the sequences.</li>
<li><strong>pad_length</strong> (int): -2 to pad according to the longest sequence, -1 to trim to the shortest sequence, any positive number for a fixed length.</li>
<li><strong>n_jobs</strong> (int): number of threads used to encode the sequences in parallel. 0 to use all CPUs available.</li>
</ul>
<h4 id="output_8">Output</h4>
<p><strong>Numpy array with shape (number of sequences, length of sequences, 3).</strong> <br /> The sequences are encoded within a cube. At each nucleotide position the Z-curve encoding gives the disparity between purines (r) and pyrimidines (y), the disparity between nucleotides with an amino (m) and a keto (k) group, and the disparity between nucleotide with weak (w) and strong (s) bonds.</p>
<pre><code class="language-python">import dna_parser as dps

sequences= [&quot;agtc&quot;,&quot;acc&quot;]
encoding= dps.zcurve_encoding(sequences)
print(encoding)
print(encoding.shape)

# Output:
#  r-y m-k w-s 
#[[[ 1  1  1]
#  [ 2  0  0]
#  [ 1 -1  1]
#  [ 0  0  0]]

# [[ 1  1  1]
#  [ 0  2  0]
#  [-1  3 -1]
#  [-1  3 -1]]]
#
#(2, 3, 3)
</code></pre>
<p><br/></p>
<h2 id="importing-sequences">Importing Sequences</h2>
<h3 id="importing-fasta-files">Importing Fasta Files</h3>
<pre><code class="language-python">dna_parser.load_fasta(path)
</code></pre>
<h4 id="function-arguments_9">Function Arguments</h4>
<ul>
<li><strong>path</strong> (str or list of str): a path or list of paths of files to import.</li>
</ul>
<h4 id="output_9">Output</h4>
<p>A list of tuples containing the metadata and sequences of each entry in the fasta file (metadata, sequence).</p>
<pre><code class="language-python">import dna_parser as dps

sequences= dps.load_fasta(&quot;path/to/fasta/file&quot;)
print(sequences)

# Output:
#[('&gt;sequence1', 'acgtatgcgtcgtc'), ('&gt;sequence2', 'cccgtga---gtcgat'), ('&gt;sequence3', 'xgtcgycaaatcg-?')]
</code></pre>
<p><br/></p>
<h3 id="importing-sequences-only">Importing Sequences Only</h3>
<pre><code class="language-python">dna_parser.load_sequences(path)
</code></pre>
<h4 id="function-arguments_10">Function Arguments</h4>
<ul>
<li><strong>path</strong> (str or list of str): a path or list of paths of files to import.</li>
</ul>
<h4 id="output_10">Output</h4>
<p>A list of str containing the sequences imported from fasta files.</p>
<pre><code class="language-python">import dna_parser as dps

sequences= dps.load_sequences(&quot;tests/seq_test.fasta&quot;)
print(sequences)

# Output:
# ['acgtatgcgtcgtc', 'cccgtga---gtcgat', 'xgtcgycaaatcg-?']
</code></pre>
<p><br/></p>
<h3 id="importing-metadata-only">Importing Metadata Only</h3>
<pre><code class="language-python">dna_parser.load_metadata(path)
</code></pre>
<h4 id="function-arguments_11">Function Arguments</h4>
<ul>
<li><strong>path</strong> (str or list of str): a path or list of paths of files to import.</li>
</ul>
<h4 id="output_11">Output</h4>
<p>A list of str containing the metadata imported from fasta files.</p>
<pre><code class="language-python">import dna_parser as dps

metadata= dps.load_metadata(&quot;tests/seq_test.fasta&quot;)
print(metadata)

# Output:
# ['&gt;sequence1', '&gt;sequence2', '&gt;sequence3']
</code></pre>
<p><br/></p>
<h2 id="other-functions">Other Functions</h2>
<h3 id="kmers">Kmers</h3>
<pre><code class="language-python">dna_parser.make_kmers(seq, k)
</code></pre>
<h4 id="function-arguments_12">Function Arguments</h4>
<ul>
<li><strong>seq</strong> (str): a genomic sequence.</li>
<li><strong>k</strong> (int): a number representing the length of kmers.</li>
</ul>
<h4 id="output_12">Output</h4>
<p>A new sequence with white spaces inserted to form kmers of length k.</p>
<pre><code class="language-python">import dna_parser as dps

kmer_seq= dps.make_kmers(&quot;agtcgtgcgtggaagagt&quot;, 3)
print(kmer_seq)

# Output:
# 'agt cgt gcg tgg aag agt '
</code></pre>
<p><br/></p>
<h3 id="generating-random-sequences">Generating Random Sequences</h3>
<pre><code class="language-python">dna_parser.random_seq(length, nb_of_seq, seq_type= &quot;dna&quot;, n_jobs= 1)
</code></pre>
<h4 id="function-arguments_13">Function Arguments</h4>
<ul>
<li><strong>length</strong> (int): length of sequences to generate.</li>
<li><strong>nb_of_seq</strong> (int): number of sequences to generate.</li>
<li><strong>seq_type</strong> (str): type of sequence to generate. either "dna", "rna", or "aa" for amino acid</li>
<li><strong>n_jobs</strong> (int): number of threads used to generate sequences in parallel.</li>
</ul>
<h4 id="output_13">Output</h4>
<p>A list of str representing the random sequences generated from a uniform probability distribution.</p>
<pre><code class="language-python">import dna_parser as dps

sequences= dps.random_seq(15,3)
print(sequences)

# Output:
# ['tagtccaaccacttg', 'gcagtactaaactca', 'caaggccatgaggta']
</code></pre>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href=".." class="btn btn-neutral float-left" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../about/" class="btn btn-neutral float-right" title="About">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href=".." style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../about/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
